vector/src/main/java/im/vector/app/features/matrixto/MatrixToBottomSheetViewModel.kt

```diff
@@ -0,0 +1,192 @@
+
+    @AssistedInject.Factory
+    interface Factory {
+        fun create(initialState: MatrixToBottomSheetState,
+                   args: MatrixToBottomSheet.MatrixToArgs): MatrixToBottomSheetViewModel
+    }
+
+    init {
+        setState {
+            copy(matrixItem = Loading())
+        }
+        viewModelScope.launch(Dispatchers.IO) {
+            resolveLink()
+        }
+    }
+
+    private suspend fun resolveLink() {
+        when {
+            args.matrixToLink != null -> {
+                val linkedId = PermalinkParser.parse(args.matrixToLink)
+                if (linkedId is PermalinkData.FallbackLink) {
+                    setState {
+                        copy(
+                                matrixItem = Fail(IllegalArgumentException(stringProvider.getString(R.string.permalink_malformed))),
+                                startChattingState = Uninitialized
+                        )
+                    }
+                    return
+                }
+
+                when (linkedId) {
+                    is PermalinkData.UserLink -> {
+                        val user = resolveUser(linkedId.userId)
+                        setState {
+                            copy(
+                                    matrixItem = Success(user.toMatrixItem()),
+                                    startChattingState = Success(Unit)
+                            )
+                        }
+                    }
+                    is PermalinkData.RoomLink -> TODO()
+                    is PermalinkData.GroupLink -> {
+                        // not yet supported
+                    }
+                    is PermalinkData.FallbackLink -> {
+                    }
+                }
+            }
+            args.userId != null       -> {
+                val user = resolveUser(args.userId)
+
+                setState {
+                    copy(
+                            matrixItem = Success(user.toMatrixItem()),
+                            startChattingState = Success(Unit)
+                    )
+                }
+            }
+            else                      -> {
+                setState {
+                    copy(
+                            matrixItem = Fail(IllegalArgumentException(stringProvider.getString(R.string.unexpected_error))),
+                            startChattingState = Uninitialized
+                    )
+                }
+            }
+        }
+    }
+
+    private suspend fun resolveUser(userId: String): User {
+        return (session.getUser(userId)
+                ?: tryOrNull {
+                    awaitCallback<JsonDict> {
+                        session.getProfile(userId, it)
+                    }
+                }?.let {
+                    User(userId, it[ProfileService.DISPLAY_NAME_KEY] as? String, it[ProfileService.AVATAR_URL_KEY] as? String)
```

Comment:
this is ugly code, we could provide something better SDK side to convert a result of getProfile() to a User